builtin/game/item_entity.lua

```diff
@@ -87,155 +78,146 @@ core.register_entity(":__builtin:item", {
-				stack:set_count(max_count)
-				count = count - max_count
-				own_stack:set_count(count)
-			else
-				self.itemstring = ''
-				stack:set_count(count)
-			end
-			local pos = object:getpos()
-			pos.y = pos.y + (count - stack:get_count()) / max_count * 0.15
-			object:moveto(pos, false)
-			local s, c
-			if not overflow then
-				obj.itemstring = stack:to_string()
-				s = 0.2 + 0.1 * (count / max_count)
-				c = s
-				object:set_properties({
-					visual_size = {x = s, y = s},
-					collisionbox = {-c, -c, -c, c, c, c},
-					wield_item = obj.itemstring
-				})
-				self.object:remove()
-				-- merging succeeded
-				return true
-			else
-				s = 0.4
-				c = 0.3
-				obj.itemstring = stack:to_string()
-				object:set_properties({
-					visual_size = {x = s, y = s},
-					collisionbox = {-c, -c, -c, c, c, c},
-					wield_item = obj.itemstring
-				})
-				s = 0.2 + 0.1 * (count / max_count)
-				c = s
-				self.itemstring = own_stack:to_string()
-				self.object:set_properties({
-					visual_size = {x = s, y = s},
-					collisionbox = {-c, -c, -c, c, c, c},
-					wield_item = self.itemstring
-				})
-			end
+	try_merge_with = function(self, own_stack, object, entity)
+		if self.age == entity.age then
+			-- Can not merge with itself
+			return false
+		end
+
+		local stack = ItemStack(entity.itemstring)
+		local name = stack:get_name()
+		if own_stack:get_name() ~= name or
+				own_stack:get_meta() ~= stack:get_meta() or
+				own_stack:get_wear() ~= stack:get_wear() or
+				own_stack:get_free_space() == 0 then
+			-- Can not merge different or full stack
+			return false
+		end
+
+		local count = own_stack:get_count()
+		local total_count = stack:get_count() + count
+		local max_count = stack:get_stack_max()
+
+		if total_count > max_count then
+			return false
 		end
-		-- merging didn't succeed
-		return false
+		-- Merge the remote stack into this one
+
+		local pos = object:get_pos()
+		pos.y = pos.y + ((total_count - count) / max_count) * 0.15
+		self.object:move_to(pos)
+
+		self.age = 0 -- Handle as new entity
+		own_stack:set_count(total_count)
+		self:set_item(own_stack)
+
+		entity.itemstring = ""
+		object:remove()
+		return true
 	end,
 
 	on_step = function(self, dtime)
 		self.age = self.age + dtime
 		if time_to_live > 0 and self.age > time_to_live then
-			self.itemstring = ''
+			self.itemstring = ""
 			self.object:remove()
 			return
 		end
-		local p = self.object:getpos()
-		p.y = p.y - 0.5
-		local node = core.get_node_or_nil(p)
-		local in_unloaded = (node == nil)
-		if in_unloaded then
-			-- Don't infinetly fall into unloaded map
-			self.object:set_velocity({x = 0, y = 0, z = 0})
+
+		local pos = self.object:get_pos()
+		local node = core.get_node_or_nil({
+			x = pos.x,
+			y = pos.y - 0.35,
+			z = pos.z
+		})
+		local vel = self.object:getvelocity()
+		local def = node and core.registered_nodes[node.name]
+		-- Ignore is nil -> stop until the block loaded
+		local is_physical = def and not def.walkable
```

Comment:
This may be just an initialised value, but physical state does not only depend on whether a walkable node is detected below as the item may not have hit it yet.